File: frontend/src/types/order.ts
export type OrderStatus =
  | 'RECEIVED'
  | 'ACCEPTED'
  | 'PREPARING'
  | 'READY_FOR_PICKUP'
  | 'OUT_FOR_DELIVERY'
  | 'DELIVERED'
  | 'REJECTED'
  | 'CANCELLED';

export interface OrderItem {
  id: number;
  order_id: number;
  product_id: number;
  product_name: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
}

export interface Order {
  id: number;
  customer_id: number;
  store_id: number;
  status: OrderStatus;
  total_amount: number;
  delivery_address: string;
  delivery_lat: number;
  delivery_lng: number;
  delivery_partner_id?: number | null;
  delivery_id?: number | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
  items: OrderItem[];
}

export interface OrderStatusHistoryItem {
  status: string;
  timestamp: string;
  note?: string | null;
}

export interface OrderStatusTracking {
  order_id: number;
  current_status: OrderStatus;
  updated_at: string;
  history: OrderStatusHistoryItem[];
}

export const ORDER_STATUS_LABELS: Record<OrderStatus, string> = {
  RECEIVED: 'New Order',
  ACCEPTED: 'Accepted',
  PREPARING: 'Preparing',
  READY_FOR_PICKUP: 'Ready for Pickup',
  OUT_FOR_DELIVERY: 'Out for Delivery',
  DELIVERED: 'Delivered',
  REJECTED: 'Rejected',
  CANCELLED: 'Cancelled',
};

export const ORDER_STATUS_COLORS: Record<OrderStatus, { bg: string; text: string; border: string }> = {
  RECEIVED: { bg: 'bg-amber-50', text: 'text-amber-700', border: 'border-amber-200' },
  ACCEPTED: { bg: 'bg-blue-50', text: 'text-blue-700', border: 'border-blue-200' },
  PREPARING: { bg: 'bg-indigo-50', text: 'text-indigo-700', border: 'border-indigo-200' },
  READY_FOR_PICKUP: { bg: 'bg-purple-50', text: 'text-purple-700', border: 'border-purple-200' },
  OUT_FOR_DELIVERY: { bg: 'bg-orange-50', text: 'text-orange-700', border: 'border-orange-200' },
  DELIVERED: { bg: 'bg-emerald-50', text: 'text-emerald-700', border: 'border-emerald-200' },
  REJECTED: { bg: 'bg-rose-50', text: 'text-rose-700', border: 'border-rose-200' },
  CANCELLED: { bg: 'bg-slate-50', text: 'text-slate-700', border: 'border-slate-200' },
};

File: backend/app/schemas/order.py
from datetime import datetime
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, Field


class OrderStatus(str, Enum):
    RECEIVED = "RECEIVED"
    ACCEPTED = "ACCEPTED"
    PREPARING = "PREPARING"
    READY_FOR_PICKUP = "READY_FOR_PICKUP"
    OUT_FOR_DELIVERY = "OUT_FOR_DELIVERY"
    DELIVERED = "DELIVERED"
    REJECTED = "REJECTED"
    CANCELLED = "CANCELLED"


# Allowed state transitions graph
VALID_ORDER_TRANSITIONS = {
    OrderStatus.RECEIVED: [OrderStatus.ACCEPTED, OrderStatus.REJECTED, OrderStatus.CANCELLED],
    OrderStatus.ACCEPTED: [OrderStatus.PREPARING, OrderStatus.CANCELLED],
    OrderStatus.PREPARING: [OrderStatus.READY_FOR_PICKUP],
    OrderStatus.READY_FOR_PICKUP: [OrderStatus.OUT_FOR_DELIVERY],
    OrderStatus.OUT_FOR_DELIVERY: [OrderStatus.DELIVERED],
    OrderStatus.DELIVERED: [],
    OrderStatus.REJECTED: [],
    OrderStatus.CANCELLED: [],
}


class OrderItemIn(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0, description="Quantity must be at least 1")


class OrderCreate(BaseModel):
    items: List[OrderItemIn] = Field(..., min_length=1, description="Order must contain at least one item")
    delivery_address: Optional[str] = Field(None, min_length=1)
    delivery_lat: Optional[float] = None
    delivery_lng: Optional[float] = None
    notes: Optional[str] = None


class OrderItemResponse(BaseModel):
    id: int
    order_id: int
    product_id: int
    product_name: str
    quantity: int
    unit_price: float
    subtotal: float


class OrderResponse(BaseModel):
    id: int
    customer_id: int
    store_id: int
    status: OrderStatus
    total_amount: float
    delivery_address: str
    delivery_lat: float
    delivery_lng: float
    delivery_partner_id: Optional[int] = None
    delivery_id: Optional[int] = None
    notes: Optional[str] = None
    created_at: str
    updated_at: str
    items: List[OrderItemResponse] = []


class OrderStatusUpdate(BaseModel):
    status: OrderStatus = Field(..., description="Target lifecycle status for order transition")


class OrderUpdate(BaseModel):
    notes: Optional[str] = None


class OrderStatusHistoryItem(BaseModel):
    status: str
    timestamp: str
    note: Optional[str] = None


class OrderStatusTrackingResponse(BaseModel):
    order_id: int
    current_status: OrderStatus
    updated_at: str
    history: List[OrderStatusHistoryItem] = []
